File: 152. Maximum Product Subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int res = INT_MIN;
        long long minProd = 1;
        long long maxProd = 1;

        for (int num : nums) {
            minProd *= num;
            maxProd *= num;
            res = max(res, (int)max(minProd, maxProd));
            if (num == 0) {
                minProd = 1;
                maxProd = 1;
            } else if (maxProd < 0) {
                maxProd = 1;
            }
        }
        minProd = 1;
        for (int i = (int)nums.size() - 1; i >= 0; i--) {
            minProd *= nums[i];
            res = max(res, (int)minProd);
            if (nums[i] == 0) minProd = 1;
        }

        return res;
    }
};

File: 152. Maximum Product Subarray.java
class Solution {
    public int calcProd(int[] nums, int negCount, int left, int right){
        if(negCount % 2 == 0){
            int prod = 0;
            for(int i = left; i < right; i++){
                if(prod == 0) prod = nums[i];
                else prod *= nums[i];
            }
            return prod;
        }

        int firstNeg = -1;
        int prod = 0;
        for(int i = left; i < right; i++){
            if(nums[i] < 0){
                negCount--;
                if(firstNeg == -1) firstNeg = i;
                if(negCount == 0) break;
            }
            if(prod == 0) prod = nums[i];
            else prod *= nums[i];
        }
        int prod2 = 0;
        for(int i = firstNeg + 1; i < right; i++){
            if(prod2 == 0) prod2 = nums[i];
            else prod2 *= nums[i];
        }

        return Math.max(prod, prod2);
    }
    public int maxProduct(int[] nums) {
        if(nums.length == 1) return nums[0];

        int negCount = 0, last = 0;
        int ans = 0;
        for(int i = 0; i < nums.length; i++){
            int num = nums[i];
            if(num < 0) negCount++;
            else if(num == 0){
                ans = Math.max(ans, calcProd(nums, negCount, last, i));
                last = i + 1;
                negCount = 0;
            }
        }
        
        ans = Math.max(ans, calcProd(nums, negCount, last, nums.length));

        return ans;
    }
}